### PR TITLE
don't add inference backedges for call sites that infer to `Any`

### PR DESCRIPTION
**Pull request by JeffBezanson**
This helps load times a decent amount. For me, `using PyCall` (second time) goes from ~20 seconds to 10 seconds. Not all packages show the same amount of improvement though.

### CODE REVIEW

**Comment by martinholters** [comment]
Removal of a method might then make call sites pessimized unnecessarily, right? Probably not a big deal, though.

File: base/compiler/abstractinterpretation.jl [hunk 1]
```diff
@@ -56,6 +56,7 @@ function abstract_call_gf_by_type(@nospecialize(f), argtypes::Vector{Any}, @nosp
     napplicable = length(applicable)
     rettype = Bottom
     edgecycle = false
+    edges = Any[]
```
Comment (vtjnash) [inline base/compiler/abstractinterpretation.jl]: fwiw, `add_backedges` is _already_ per-callsite

**Comment by JeffBezanson** [comment]
That's true. But, I believe the most important cases are when imprecision comes from the argument types, not from the called methods. For example, we often deliberately specialize a function for `Any` arguments and reuse it to save compile time. Inside such a function, there are lots of call sites like `f(x::Any)::Any`. Adding an edge from `f` partly defeats the purpose of such generic code.

So we could potentially improve this by taking the argument types into account --- e.g. if the argument types are pretty tight but we inferred `Any`, still add a backedge in case we can get better type information later when more methods are added or removed.

File: base/compiler/abstractinterpretation.jl [hunk 2]
```diff
@@ -297,18 +307,16 @@ function abstract_call_method(method::Method, @nospecialize(sig), sparams::Simpl
...
+            return Any, false, nothing
         end
         sparams = recomputed[2]::SimpleVector
     end
 
     rt, edge = typeinf_edge(method, sig, sparams, sv)
     if edge === nothing
         edgecycle = true
-    else
-        add_backedge!(edge::MethodInstance, sv)
```
Comment (vtjnash) [inline base/compiler/abstractinterpretation.jl]: if `edgecycle` is true, this is the wrong edge to create, but it also means that we will compute and add the real edge later

edit: theoretically, we should be able to see the `Any` information result and use that to break the callgraph cycle on this edge here (possibly permitting faster convergence), but that does weird things to the convergence algorithm postulates

**Comment by vtjnash** [comment]
> might then make call sites pessimized unnecessarily

Also adding methods, but we've already previously decided (de-facto) that we generally don't care to track changes to the system that might affect precision, only correctness

File: base/compiler/abstractinterpretation.jl [hunk 3]
```diff
@@ -67,14 +68,20 @@ function abstract_call_gf_by_type(@nospecialize(f), argtypes::Vector{Any}, @nosp
...
                 rettype = tmerge(rettype, rt)
                 rettype === Any && break
             end
             rettype === Any && break
         else
-            rt, edgecycle = abstract_call_method(method, sig, match[2]::SimpleVector, sv)
+            rt, edgecycle, edge = abstract_call_method(method, sig, match[2]::SimpleVector, sv)
+                if edge !== nothing
+                    push!(edges, edge)
+                end
```
Comment (vtjnash) [inline base/compiler/abstractinterpretation.jl]: indent

**Comment by ufechner7** [comment]
Any idea why three of the checks are failing?

**Comment by JeffBezanson** [comment]
Looks like a problem with httpbin.

**Pull request merged**